infra/src/main/resources/db/migration/prod/V86__relax_design_row_id_uniqueness_constraints.sql

```diff
@@ -0,0 +1,29 @@
+alter table layout.track_number
+  drop constraint track_number_design_row_id_unique,
+  add constraint track_number_design_row_id_unique_in_designs exclude (design_row_id with =) where (design_row_id is not null and design_id is not null),
+  add constraint track_number_design_row_id_unique_in_main exclude (design_row_id with =) where (design_row_id is not null and design_id is null),
+  add constraint track_number_only_draft_has_design_row check (draft or design_row_id is null);
```

Comment:
Estääkös tämä nyt suunnitelman sisäisen julkaisun, jos `design_row_id is not null and draft = false` ei ole sallittu kombinaatio? (Vai luenko tätä SQL:ää vain jotenkin väärin? Tavallisesti olen tottunut siihen, että kaikissa boolean-operaatioissakin on varsinainen operaatio avattuna kokonaisuudessaan, ei shorthandeja) 🤔

Reply:
Juuei: Design-draftilla on aina design_row olemassa, design-officialilla se on aina null.

Itse asiassa pyöräytänpä taskin vielä siitä, että tarkistaa että pitäisikö tehdä noita check-constraintteja vielä lisääkin, tuossahan kun ei vielä ole kaikkia rajoitteita, mitä kantaan voisi määrittää (esim. että main-officialilla ei saa olla design_row_id:tä eikä official_row_id:tä)

Reply:
Aa niin aivan. Tuon columnin nimeäminen on sitten outo (sen voisi parantaa, mutta se nyt ei ole suoraan tämän tiketin asiaa.) Nykyisen nimen perusteella olettaisin että kaikilla jossain suunnitelmissa tehdyillä asioilla on `design_row_id` riippumatta onko ne siellä sisäisesti draft vai official

Reply:
hmm, niin no kaikki nuo x_row_id ovat "se toinen ilmentymä tästä asiasta" ja siten design_row_id on siis "tämän käsitteen design-versio". Jos kyseessä on official design, niin eihän se voisi mihinkään toiseen design-riviin viitata eikä omaa id:tä ole järkeä tuohon toiseen viittaus-columniin kopioida. Tuo että milloin sen pitää olla null ja milloin ei, on kyllä vähän monimutkainen ja tarkkuutta vaativa mutta on se sentään dokumentoitu feature. Toki parempia nimiä voi silti ehdottaa.

Reply:
Mitä tuo `(design_row_id with =)` syntaksi tarkoittaa? Itse rajoitus näyttää järkevältä.

Reply:
Se on exclude-rajoitteiden syntaksi: `<lause> with <operaattori>`, millä postgres sitten estää taulussa olemasta kahta riviä, joilla lasketuilla lauseilla operaattorin arvo on true. Tässä tapauksessa kun käytetään pelkkää yhtäsuuruusoperaattoria, saisi aivan saman asian aikaan myös unique indexillä, mutta exclude-rajoitteella asian saa sanottua juuri rajoitteena.